buildkite.rb

```diff
@@ -1,9 +1,75 @@
 #!/usr/bin/env ruby
 
-git_output = `git log -m -1 --name-only --pretty=format:%n`
+require 'json'
+require 'net/http'
+require 'uri'
+
+BUILDKITE_ACCESS_TOKEN = ENV['BUILDKITE_ACCESS_TOKEN']
+BUILDKITE_PROJECT = ENV['BUILDKITE_PROJECT'] ||= 'bento'
+BUILDKITE_ORGANIZATION = ENV['BUILDKITE_ORGANIZATION_SLUG'] ||= 'chef'
+
+def buildkite_api_uri(args = {})
+  raise Exception.new('Missing project argument') if args[:project].nil?
+  raise Exception.new('Missing endpoint argument') if args[:endpoint].nil?
+
+  URI::HTTPS.build(
+    host:   'api.buildkite.com',
+    path:   "/v2/organizations/#{BUILDKITE_ORGANIZATION}/pipelines/#{args[:project]}/#{args[:endpoint]}",
+    query:  "access_token=#{BUILDKITE_ACCESS_TOKEN}"
+  )
+end
+
+# Returns an array of builds hashes for the environment defined project.
+def buildkite_builds
+  response = Net::HTTP.get_response(
+      buildkite_api_uri(project: BUILDKITE_PROJECT, endpoint: 'builds')
+  )
+
+  if response.code.to_i > 200
+    raise Exception.new("Unexpected response from BuildKite API: #{response.code_type}")
+  else
+    JSON.parse(response.body)
+  end
+end
+
+# Returns a string 
+def last_passed_build_git_hash
+  git_hash = ''
+
+  buildkite_builds.map do |build|
+    if build['state'] == 'passed'
+      git_hash = build['commit']
+      break
+    end
+  end
+
+  if git_hash.empty?
+    raise Exception.new('Unable to determine the commit hash of the last successful build.')
+  else
+    git_hash
+  end
+end
+
+
+# Return an array of files changed since the last successful build.
+# Sorry for the long function name.
+def changed_files_since_last_passed_build
+  # `rev-list commit_x..HEAD` shows the commit hashes for each commit from a
+  # given commit to the current HEAD.
+  # `--objects` shows the actual files which have changed in the series of
+  # commits.
+  # We strip the commit IDs via awk because its cheap and a lazy way to get only
+  # the data back we need.
+
+  objects = `git rev-list #{last_passed_build_git_hash}..HEAD  --objects | awk '{print $2}'`.split.uniq
+  # Only return objects which are files, or files with a path.
+  objects.select { |object| object if File.file?(object) }
+end
 
 # TODO: decide what, if anything, is rebuilt if bin/bento, buildkite.rb, scripts/common change.
-changed_files = git_output.split("\n").reject { |t| t.empty? || t.include?('gitignore') || t.include?('dummy_metadata') || t.include?('bin/bento') || t.include?('buildkite.rb') || t.include?('scripts/common') || t.include?('.md')}
+# This expects the input array to only contain files and paths to files.
+# eg: [".travis.yml", "CHANGELOG.md","README.md", "debian-8.3-amd64.json"]
+changed_files = changed_files_since_last_passed_build { |t| t.empty? || t.include?('gitignore') || t.include?('dummy_metadata') || t.include?('bin/bento') || t.include?('buildkite.rb') || t.include?('scripts/common') || t.include?('.md')}
```

Comment:
Assuming you make the change I suggested in https://github.com/chef/bento/pull/553/files#r57823160 you could just use `changed_files_since_last_passed_build ` in the rest of the file...no need for this extra `changed_files` variable.